rfc/system/3844-artifact-service.md

```diff
@@ -0,0 +1,148 @@
+
+#### Create a new one locally
+Use case: I have some data locally, I want to upload this to Flyte and make the Artifact service aware of this, as if it were the output of a workflow execution.
+
+```python
+a = Artifact.new(python_val, pytype, schema, metadata, ...)  # remote detected from current context maybe.
+a = Artifact.new(python_val, pytype, schema, metadata, remote, type=Dataset ...)
+new_a = Artifact.new(large_important_str)  # .new(my_df), .new(Path("/root/data/tokened/"))
+remote.create(new_a)
+```
+
+Given an object in Python memory, whether it's a float, a pd.DataFrame, or a io.StringIO buffer, a user should be able to turn this into an Artifact. How this might work is:
+* FlyteRemote adds itself into the data persistence layer as an fsspec compatible filesystem by adding a new context that overrides the output data prefix with `flyte://dv1/proj/domain/`. See this [ongoing work](https://github.com/flyteorg/flytekit/pull/1674/files).
+* Run the Python object through the flytekit Type Engine.
+* Return the Artifact object containing a possibly different flyte URL to the user. Attributes of the created Artifact might contain
+  * an address that's universal per Artifact service (aka Flyte installaion).
+  * a name that was autogenerated using the item stored.
+  * aliases that make sense that are autogenerated
+  * it will be associated with a Literal and a Literal type and also store any schema or metadata information.
+  * bookkeeping information, created_at, user who created it, etc.
+
+#### Queryable and bindable queries
+Use case: Identify the artifact with such and such name and the "latest" alias. I have a workflow whose default input I want to be whatever is returned by that search when the workflow is kicked off.
+
+Basic functionality here on the search side might look like
+* Given an Artifact URL, list all executions it was used in.
+* Given an Artifact URL, find the execution that produced it, or let the user know if it was an upload.
+* Given a set of aliases or tags, return Artifact objects.
+
+On the authoring side, we would like users to be able to do the following
+```python
+@task
+def t1() -> Annotated[nn.Module, Artifact(name="models.nn.lidar", alias="latest", overwrite_alias=True)]: ...
+
+# Possibly in a completely different repo
+@workflow
+def wf(model: nn.Module = Artifact.get_query(name="models.nn.lidar", alias="latest")): ...
+```
+The binding is made at compilation time but it is not resolved until run time. That is, when `wf` is run, Admin will have to call out to the Artifact service, get back precisely one Artifact, and use the Literal therein contained as the input. If the type is not a match, then it can result it in a user run time error, which is okay.
+
+## 3 Proposed Implementation
```

Comment:
Use case:
As a user I would like to visualize the lineage graph of all these artifacts, links being created by a task that has an artifact as input and producing another artifact.

I don't think Flyte would have to implement this visualization itself but there could be integrations with e.g. https://openlineage.io later on.